### PR TITLE
Keyboard shortcuts for track construction

### PR DESCRIPTION
**Pull request by JoelTroch**
This PR add some keyboard shortcuts for track construction (especially for custom designs). I tried my best to optimize and have a clean code as much as possible.

Related issue: #1730

### CODE REVIEW

File: src/openrct2/interface/keyboard_shortcut.c [hunk 1]
```diff
@@ -635,6 +636,424 @@ static void shortcut_debug_paint_toggle()
 	}
 }
 
+static void shortcut_ride_construction_turn_left()
+{
+	if (gScreenFlags & SCREEN_FLAGS_TITLE_DEMO)
+		return;
+
+	rct_window * w = window_find_by_class(WC_RIDE_CONSTRUCTION);
```
Comment (marijnvdwerf) [inline src/openrct2/interface/keyboard_shortcut.c]: I'd rather see this code in the construction window file.

Comment (marijnvdwerf) [inline src/openrct2/interface/keyboard_shortcut.c]: (Reason being that the `widgets[9]` etc is difficult to maintain)

Comment (JoelTroch) [inline src/openrct2/interface/keyboard_shortcut.c]: You mean move the whole code to `src/openrct2/windows/ride_construction.c` ?

**Comment by JoelTroch** [comment]
I have refactored the "turn left" keyboard shortcut, can you please tell me if it's better that way ? If it's good then I'll do the same for the rest

File: src/openrct2/ride/track.h [hunk 2]
```diff
@@ -537,4 +537,6 @@ void game_command_remove_track(sint32 *eax, sint32 *ebx, sint32 *ecx, sint32 *ed
 void game_command_set_maze_track(sint32 *eax, sint32 *ebx, sint32 *ecx, sint32 *edx, sint32 *esi, sint32 *edi, sint32 *ebp);
 void game_command_set_brakes_speed(sint32 *eax, sint32 *ebx, sint32 *ecx, sint32 *edx, sint32 *esi, sint32 *edi, sint32 *ebp);
 
+void window_ride_construction_keyboard_shortcut_turn_left();
```
Comment (marijnvdwerf) [inline src/openrct2/ride/track.h]: This should be in `window.h`

File: src/openrct2/windows/ride_construction.c [hunk 3]
```diff
@@ -4060,3 +4060,126 @@ void window_ride_construction_do_entrance_exit_check()
...
 }
+
+void window_ride_construction_keyboard_shortcut_turn_left()
+{
+	rct_window *w = window_find_by_class(WC_RIDE_CONSTRUCTION);
+	if (w == NULL || widget_is_disabled(w, WIDX_STRAIGHT) || w->widgets[WIDX_STRAIGHT].type == WWT_EMPTY) {
+		return;
+	}
+
+	switch (_currentTrackCurve) {
```
Comment (marijnvdwerf) [inline src/openrct2/windows/ride_construction.c]: Indentation of the switch cases is inconsistent.

**Comment by marijnvdwerf** [comment]
Made a few style-related comments.

@IntelOrca Can you go over `window_ride_construction_keyboard_shortcut_turn_left` implementation-wise?

**Comment by JoelTroch** [comment]
Methods declaration moved from `track.h` to `window.h` and fixed the bad coding style for `switch` statements.

I'm doing the rest while waiting for @IntelOrca 's review implementation-wise.

**Comment by Nubbie** [comment]
Maybe could squash them all together in the end and insert a 'Fix: XXXX'? :p

**Comment by JoelTroch** [comment]
@Nubbie Commits have been squashed ^^

**Comment by Broxzier** [comment]
My laptop has a Numpad that I have disabled most of the time, with some of the keys having double meaning (8/up, 4/left, 6/right, 2/down, 7/home, 9/page up, 1/end, 3/page down), but this doesn't take into account whether I have Numpad enabled or disabled. With Numpad enabled I expect Num 9 to act only as Num 9, and with Numpad disabled I expect it to act only act as the page up key (in my case of course).

**Comment by JoelTroch** [comment]
If I understood correctly, you want "Keypad 8" (I'm using the term "Keypad" because that's how SDL2 call them) and "Page Up" to act the same ?

**Comment by Broxzier** [comment]
@JoelTroch No, exactly the opposite.

img_20170217_150620201

In other applications when I have Numpad disabled (Numlock off), Keypad 9 acts like Page Up, and when Numpad is enabled, it acts as "Keypad 9" (In a text editor, it would either scroll up, or write down the "9").
In OpenRCT2 however, it always acts as if Numpad is disabled, and zooms out when I press "Keypad 9", even when Numpad is enabled.

This issue has probably been here before this PR too, I just only found out about it now.

**Comment by marijnvdwerf** [comment]
This issue should probably get solved by #3908. Unfortunately that PR is rather stuck.

**Comment by JoelTroch** [comment]
@Broxzier I see what you mean, I have the same layout as you.

After a very quick Google search, it might be related to Unicode input but someone more expert with input stuff should confirm or deny my sentence.

**Comment by janisozaur** [comment]
What's the status of this?

**Comment by JoelTroch** [comment]
It can be merged whenever the main OpenRCT2 developers desire to do so.

About the issue mentioned by @Broxzier , this isn't in the scope of this PR but @marijnvdwerf mentioned that #3908 could fix it if it's rebased and free of merge conflicts.

**Comment by janisozaur** [comment]
@JoelTroch can you rebase?

**Comment by JoelTroch** [comment]
@janisozaur Done ^^

File: contributors.md [hunk 4]
```diff
@@ -55,6 +55,7 @@ Includes all git commit authors. Aliases are GitHub user names.
 * Niels NTG Poldervaart (Niels-NTG) - Misc.
 * (zaxcav) - Improvements to original pathfinding algorithm.
 * Hielke Morsink (Broxzier) - Tile inspector, misc.
+* Joël Troch (Shepard62FR) - Keyboard shortcuts for ride construction.
```
Comment (Gymnasiast) [inline contributors.md]: Your Github username is JoelTroch, not Shepard62FR.

File: distribution/changelog.txt [hunk 5]
```diff
@@ -1,5 +1,6 @@
 0.0.7 (in development)
 ------------------------------------------------------------------------
+- Feature: [#1730] Add keyboard shortcuts for track construction.
```
Comment (Gymnasiast) [inline distribution/changelog.txt]: `- Feature: [#1730] Keyboard shortcuts for track construction.`

**Comment by JoelTroch** [comment]
Requested changes applied

**Comment by janisozaur** [comment]
@Gymnasiast I looked briefly at the code and it looked simple enough for me not to expect any bugs in there. There was still a misindented line or two there. I trust your judgement.

**Comment by JoelTroch** [comment]
@janisozaur I fixed one of the misaligned line, let me know if there are more

**Pull request merged**